fdbserver/Status.actor.cpp

```diff
@@ -1028,15 +1028,31 @@ ACTOR static Future<JsonBuilderObject> recoveryStateStatusFetcher(WorkerDetails
 	state JsonBuilderObject message;
 
 	try {
-		state Future<TraceEventFields> activeGens = timeoutError(mWorker.interf.eventLogRequest.getReply( EventLogRequest( LiteralStringRef("MasterRecoveryGenerations") ) ), 1.0);
-		TraceEventFields md = wait( timeoutError(mWorker.interf.eventLogRequest.getReply( EventLogRequest( LiteralStringRef("MasterRecoveryState") ) ), 1.0) );
+		std::vector<Future<TraceEventFields>> futures;
+		futures.push_back(timeoutError(mWorker.interf.eventLogRequest.getReply( EventLogRequest( LiteralStringRef("MasterRecoveryGenerations") ) ), 1.0));
+		futures.push_back(timeoutError(mWorker.interf.eventLogRequest.getReply( EventLogRequest( LiteralStringRef("MasterRecoveryFullyRecovered") ) ), 1.0));
+		futures.push_back(timeoutError(mWorker.interf.eventLogRequest.getReply( EventLogRequest( LiteralStringRef("MasterRecoveryState") ) ), 1.0));
+		std::vector<TraceEventFields> msgs = wait(getAll(futures));
+
+		const TraceEventFields& md = msgs[2];
 		int mStatusCode = md.getInt("StatusCode");
 		if (mStatusCode < 0 || mStatusCode >= RecoveryStatus::END)
 			throw attribute_not_found();
 
 		message = JsonString::makeMessage(RecoveryStatus::names[mStatusCode], RecoveryStatus::descriptions[mStatusCode]);
 		*statusCode = mStatusCode;
 
+		const TraceEventFields& mLastRecoveryMsg = msgs[1];
+		std::string lastFullyRecoveredTimeS;
+		if (mLastRecoveryMsg.tryGetValue("Time", lastFullyRecoveredTimeS)) {
+			double lastFullyRecoveredTime = atof(lastFullyRecoveredTimeS.c_str());
+			// `lastFullyRecoveredTime` is the timestamp taken on master so the time interval calculated below may not
+			// be accurate due to the clock skew across the network, but it's good enough for the purpose it's used.
+			message["time_since_last_fully_recovered_seconds"] = now() - lastFullyRecoveredTime;
```

Comment:
You could potentially use versions here to measure time, depending in part on whether we are intending to only view the most recent recovery.

Reply:
I assumed clock skew won't be so big that it even matters in this use case. @brownleej How accurate does this time need to be?

Reply:
There are good reasons to keep clock skew low and in practice it may not be that important to us specifically, but in general we try to minimize situations where we rely on the accuracy of clocks. Having this field be inaccurate may not be hugely impactful, but at the same time it's probably not too hard to make it better.

Reply:
Done.